### PR TITLE
support ST_Contains with H3 optimization

### PR DESCRIPTION
**Pull request by vishwa35**
## Description

Adds a second H3IndexFilterOperator specifically for inclusions; the former was designed for distance/radius calculations.

As a part of this change, a number of functions and logic to estimate a resolution were added to the ScalarFunctions. A hardcoded value currently exists in L71 of H3InclusionIndexFilterOperator that should probably we sorted out.

It also enables ST_Contains on geography coordinates, as there isn't anything different about that from geometry that should prohibit that? The geo coords are isomorphic wrt cartesian coords.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java [hunk 1]
```diff
@@ -83,9 +82,6 @@ public TransformResultMetadata getResultMetadata() {
     for (int i = 0; i < projectionBlock.getNumDocs(); i++) {
       Geometry firstGeometry = GeometrySerializer.deserialize(firstValues[i]);
       Geometry secondGeometry = GeometrySerializer.deserialize(secondValues[i]);
-      if (GeometryUtils.isGeography(firstGeometry) || GeometryUtils.isGeography(secondGeometry)) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java:86]: why did you remove this? performance?

**Review by kishoreg — COMMENTED** [review]
@vishwa35 thank you for submitting this PR. Can you please add some test cases and also list sample queries in the PR description. We will also need to update the docs.

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java [hunk 2]
```diff
@@ -111,4 +138,35 @@ public static String stAsText(byte[] bytes) {
...
+
+    for (int i = 0; i < n; i++) {
+      for (int j = i + 1; j < n; j++) {
+        max = Math.max(max, dist(points.get(i), points.get(j)));
+      }
+    }
+    return Math.sqrt(max);
+  }
+
+  public static double dist(Coordinate a, Coordinate b) {
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/ScalarFunctions.java:169]: either make the unit an argument or  add the unit as part of the func name

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java (hunk 1 above)
Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java:86]: this check is needed, the current implementation does not handle geography correctly.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.pinot.core.operator.docidsets.BitmapDocIdSet;
+import org.apache.pinot.segment.local.utils.GeometrySerializer;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.H3IndexReader;
+import org.apache.pinot.spi.utils.BytesUtils;
+import org.locationtech.jts.geom.Coordinate;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+/**
+ * A filter operator that uses H3 index for geospatial data inclusion
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java:36]: plz add comments about the algorithm used in this operator

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                ScalarFunctions.calcResFromMaxDist(ScalarFunctions.maxDist(coordinateList), 50));
+    }
+
+    @Override
+    protected FilterBlock getNextBlock() {
+        // have list of h3 hashes for polygon provided
+        // return filtered num_docs
+        MutableRoaringBitmap fullMatchDocIds = new MutableRoaringBitmap();
+        for (long docId : _h3Ids) {
+            fullMatchDocIds.or(_h3IndexReader.getDocIds(docId));
```
Comment (yupeng9) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java:80]: how do you determine full matches?

Comment (yupeng9) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java:80]: please add unit tests to verify the algorithm

**Review by yupeng9 — CHANGES_REQUESTED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java (hunk 1 above)
Comment (vishwa35) [reply pinot-core/src/main/java/org/apache/pinot/core/geospatial/transform/function/StContainsFunction.java:86]: why does it not work for geography?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+import org.apache.pinot.segment.spi.index.reader.H3IndexReader;
+import org.apache.pinot.spi.utils.BytesUtils;
+import org.locationtech.jts.geom.Coordinate;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+/**
+ * A filter operator that uses H3 index for geospatial data inclusion
+ */
+public class H3InclusionIndexFilterOperator extends BaseFilterOperator {
+    private static final String OPERATOR_NAME = "H3IndexFilterOperator";
```
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java:39]: make operator name consistent with the class name?
